mail/api.py

```diff
@@ -322,18 +321,26 @@ def add_automatic_email(original_search, email_subject, email_body, sender_name,
         )
 
 
-@transaction.atomic
+@contextmanager
 def mark_emails_as_sent(automatic_email, emails):
     """
-    Mark users who have the given emails as sent
+    Context manager to mark users who have the given emails as sent after successful sending of email
 
     Args:
         automatic_email (AutomaticEmail): An instance of AutomaticEmail
         emails (iterable): An iterable of emails
     """
-    users = User.objects.filter(email__in=emails)
-    for user in users:
-        SentAutomaticEmail.objects.get_or_create(
-            user=user,
-            automatic_email=automatic_email,
-        )
+    with transaction.atomic():
+        user_ids = User.objects.filter(email__in=emails).exclude(
+            sentautomaticemail__automatic_email=automatic_email,
+        ).values_list('id', flat=True)
+
+        for user_id in user_ids:
+            SentAutomaticEmail.objects.create(
```

Comment:
I think you should move this out of the transaction and switch back to `get_or_create` because a) you only want to send emails that you have an atomic guarantee you were the only one who did so. The best way I know to do that is to create them outside the transaction so they get committed and then only `select_for_update` and yield the emails for those which you actually created and then pass that into `send_batch`. Otherwise, you could yield down below and then the entire transaction fails to commit because you couldn't create one record or some other race condition.